src/verify.c

```diff
@@ -0,0 +1,182 @@
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  if (0 == sk_X509_push(certs, cert)) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  // NOVERIFY is set to avoid validating the certificate chain for signing.
+  // Since the signatures this library is designed to verify will always be
+  // self-signed, the NOVERIFY option is required for the verification to work
+  if (1 == PKCS7_verify(p7, certs, store, bio_document, NULL,
+                        PKCS7_NOINTERN | PKCS7_NOVERIFY)) {
+    rv = VF_SUCCESS;
+  } else {
+    // The last error in the error queue ought to be a "signature verification"
+    // error as that's the last error to emit when a document validates to
+    // invalid.  In order to determine whether to treat this as VF_FAIL or
+    // VF_EXCEPTION, first determine whether this error is a signature
+    // verification error or not.  If it is, mark this invocation of VF_verify
+    // as VF_FAIL, then clear the error queue so that if future errors occur,
+    // handle them as exceptions.  This also ensures that exceptions during
+    // verification are handled differently to invalid signatures
+    unsigned long errorNum = ERR_peek_last_error();
+
+    if (ERR_GET_LIB(errorNum) == ERR_LIB_PKCS7 &&
+        ERR_GET_FUNC(errorNum) == PKCS7_F_PKCS7_VERIFY &&
+        ERR_GET_REASON(errorNum) == PKCS7_R_SIGNATURE_FAILURE) {
+      ERR_clear_error();
+      rv = VF_FAIL;
+    } else {
+      rv = VF_EXCEPTION;
+    }
+  }
+
+end:
+  if (!BIO_free(bio_document) || !BIO_free(bio_pubkey) ||
+      !BIO_free(bio_pkcs7)) {
+    rv = VF_EXCEPTION;
+  }
+
+  PKCS7_free(p7);
+  X509_STORE_free(store);
+  sk_X509_free(certs);
+  X509_free(cert);
+
+  struct Error *head = NULL;
+
+  unsigned long errorNum = ERR_peek_error();
+
+  if (!errorNum && rv == VF_EXCEPTION) {
+    // This case is for there being an exception signaled in this file but
+    // there isn't a corresponding error in the OpenSSL error queue.  Ideally,
+    // we'd use ERR_put_error to insert error messages which we could use to
+    // display using a single error reporting system.
+    head = malloc(sizeof(struct Error));
+    head->reason = "IID-Verify Exception";
+    head->lib = __FILE__;
+    head->func = "VF_verify";
+    head->next = NULL;
+    head->line = __LINE__;
+    *err = head;
+  } else if (errorNum && err == NULL) {
+    rv = VF_EXCEPTION;
+  } else if (errorNum && err != NULL) {
+    rv = VF_EXCEPTION;
+    do {
+      struct Error *new = malloc(sizeof(struct Error));
+      errorNum = ERR_get_error_line(&new->file, &new->line);
+      // This break should *not* be needed because before this invocation, only
+      // the ERR_peek_error() function has been called at that point.  There
+      // shouldn't have been any removals from the error queue.  Sadly, this is
+      // not the case, since the call to ERR_get_error_line here returns a
+      // falsy value, which means that the error queue doesn't work as
+      // documented in the man page.
```

Comment:
Maybe I'm misunderstanding this comment, but the behavior makes sense to me.  Assuming there was only one error in the queue, on the first iteration of the `do-while`, this call will pop the entry from the queue.  On the second iteration, it will return 0.  Continuing on to call `ERR_reason_error_string` for that 0 would be an error, so you need to exit the loop here (and the `while (errNum)` at the end is redundant.

Reply:
I'll modify it to a `for(;;)` loop.  I think I'd been looking at the code too long when I wrote that.  I thought it was an error in the interaction between the previous peek() call and this get() call, but you're right here.